src/components/ActionDetail/ActionDetail.tsx

```diff
@@ -37,50 +37,59 @@ import { FormControl } from './FormControl';
 import { ActionContext } from '../../../common/types';
 import { actionTitle } from '../../common/shared';
 
-function createUpdatedAction(field: string, value: string, context: ActionContext) {
+function createUpdatedAction(fields: IActionUpdateFields, context: ActionContext) {
+  const action = {
+    ...context.action,
+    url: fields.url,
+    selector: fields.selector,
+    text: fields.text,
+  };
   const actionContext = {
     ...context,
-    action: { ...context.action, [field]: value },
+    action,
     modified: true,
     isOpen: false,
   };
   const title = actionTitle(actionContext.action);
   return title ? { ...actionContext, title } : actionContext;
 }
-
+interface IActionUpdateFields {
+  url: string;
+  selector: string;
+  text?: string;
+}
 interface IActionDetail {
   actionContext: ActionContext;
   actionIndex: number;
   stepIndex: number;
   close?: () => void;
+  update?: (action: ActionContext, stepIndex: number, actionIndex: number) => void;
```

Comment:
Where is this prop actually supplied to `ActionDetail`? It seems the only place where we're applying this prop is in the tests, unless I'm missing something?

Reply:
Yes, Actually this is the part I wasn't really happy with.. I tried to spy on the `onUpdateAction` but didn't want to introduce the `StepContext` for the unit tests so I ended up adding the prop that is only used by tests. Do you have any suggestions?

Reply:
Actually, I've updated the tests to use StepContext so that we didn't need to add a new prop `update` just for the test purpose. Thanks for bringing this to my attention.

Reply:
A common pattern, too, is to overload the RTL `render` function with a custom test helper that will wrap the components with some custom boilerplate and remove that burden from the test files.

I think we likely even have one of those helpers in this project already. I can check for you in a bit.